tensorflow/lite/micro/micro_profiler.h

```diff
@@ -33,67 +46,185 @@ class MicroProfiler : public MicroProfilerInterface {
+          "MicroProfiler errored out because total number of events exceeded "
+          "the maximum of %d.",
+          MAX_EVENTS);
+      TFLITE_ASSERT_FALSE;
+    }
+
+    tags_[num_events_] = tag;
+    start_ticks_[num_events_] = GetCurrentTimeTicks();
+    end_ticks_[num_events_] = start_ticks_[num_events_] - 1;
+    return num_events_++;
+  }
 
   // Marks the end of an event associated with event_handle. It is the
   // responsibility of the caller to ensure than EndEvent is called once and
   // only once per event_handle.
   //
   // If EndEvent is called more than once for the same event_handle, the last
-  // call will be used as the end of event marker. If EndEvent is called 0 times
+  // call will be used as the end of event marker.If EndEvent is called 0 times
   // for a particular event_handle, the duration of that event will be 0 ticks.
-  virtual void EndEvent(uint32_t event_handle) override;
+  void EndEvent(uint32_t event_handle) {
+    TFLITE_DCHECK(event_handle < MAX_EVENTS);
+    end_ticks_[event_handle] = GetCurrentTimeTicks();
+  }
 
   // Clears all the events that have been currently profiled.
-  void ClearEvents() { num_events_ = 0; }
+  void ClearEvents() {
+    num_events_ = 0;
+    num_tag_groups_ = 0;
+  }
 
   // Returns the sum of the ticks taken across all the events. This number
   // is only meaningful if all of the events are disjoint (the end time of
   // event[i] <= start time of event[i+1]).
-  uint32_t GetTotalTicks() const;
+  uint32_t GetTotalTicks() const {
+    int32_t ticks = 0;
+    for (int i = 0; i < num_events_; ++i) {
+      ticks += end_ticks_[i] - start_ticks_[i];
+    }
+    return ticks;
+  }
 
   // Prints the profiling information of each of the events in human readable
   // form.
-  void Log() const;
+  void Log(MicroProfilerLogFormat format) const {
+#if !defined(TF_LITE_STRIP_ERROR_STRINGS)
+    switch (format) {
+      case MicroProfilerLogFormat::HumanReadable:
+        for (int i = 0; i < num_events_; ++i) {
+          uint32_t ticks = end_ticks_[i] - start_ticks_[i];
+          uint64_t us = TicksToUs(ticks);
+          MicroPrintf("%s took %" PRIu64 ".%" PRIu64 " ms (%u ticks)", tags_[i],
+                      us / 1000, us % 1000, ticks);
+        }
+        break;
 
-  // Prints the profiling information of each of the events in CSV (Comma
-  // Separated Value) form.
-  void LogCsv() const;
+      case MicroProfilerLogFormat::Csv:
+        MicroPrintf("\"Event\",\"Tag\",\"Ms\",\"Ticks\"");
+        for (int i = 0; i < num_events_; ++i) {
+#if defined(HEXAGON) || defined(CMSIS_NN)
+          int ticks = end_ticks_[i] - start_ticks_[i];
+          MicroPrintf("%d,%s,%u,%d", i, tags_[i], TicksToMs(ticks), ticks);
+#else
+          uint32_t ticks = end_ticks_[i] - start_ticks_[i];
+          uint64_t us = TicksToUs(ticks);
+          MicroPrintf("%d,%s,%" PRIu64 ".%" PRIu64 ",%" PRIu32, i, tags_[i],
+                      us / 1000, us % 1000, ticks);
+#endif
```

Comment:
Changing the CSV format probably won't pass review, as it most likely breaks countless internal Google, and external third-party scripts.

Reply:
Removed the `Ms` column